iothub_client/src/iothub_client_diagnostic.c

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdlib.h>
+#include <math.h>
+#include "azure_c_shared_utility/optimize_size.h"
+#include "azure_c_shared_utility/gballoc.h"
+#include "azure_c_shared_utility/xlogging.h"
+#include "azure_c_shared_utility/buffer_.h"
+
+#include "iothub_client_diagnostic.h"
+
+static const int BASE_36 = 36;
+
+static char* get_current_time_utc(char* timeBuffer, int bufferLen, const char* formatString)
+{
+    time_t epochTime;
+    struct tm* utcTime;
+    epochTime = time(NULL);
+    utcTime = gmtime(&epochTime);
+    strftime(timeBuffer, bufferLen, formatString, utcTime);
+    return timeBuffer;
+}
+
+static char get_base36_char(unsigned char value)
+{
+    if (value <= 9)
+    {
+        return '0' + value;
+    }
+    else
+    {
+        return 'a' + value - 10;
+    }
+}
+
+static char* generate_eight_random_characters(char *randomString)
+{
+    char* randomStringPos = randomString;
+    for (int i = 0; i < 4; ++i)
+    {
+        int rawRandom = rand();
+        int first = rawRandom % BASE_36;
+        int second = rawRandom / BASE_36 % BASE_36;
+        *randomStringPos++ = get_base36_char((unsigned char)first);
+        *randomStringPos++ = get_base36_char((unsigned char)second);
+    }
+    *randomStringPos = 0;
+
+    return randomString;
+}
+
+static bool should_add_diagnostic_info(IOTHUB_DIAGNOSTIC_SETTING_DATA* diagSetting)
+{
+    if (diagSetting->diagSamplingPercentage > 0)
+    {
+        if (diagSetting->currentMessageNumber == UINT32_MAX)
+        {
+            diagSetting->currentMessageNumber %= diagSetting->diagSamplingPercentage;
+        }
+        ++diagSetting->currentMessageNumber;
+
+        double number = diagSetting->currentMessageNumber;
+        double percentage = diagSetting->diagSamplingPercentage;
+        return (floor((number - 2) * percentage / 100.0) < floor((number - 1) * percentage / 100.0));
+    }
+    return false;
+}
+
+bool IoTHubClient_Diagnostic_AddIfNecessary(IOTHUB_DIAGNOSTIC_SETTING_DATA* diagSetting, IOTHUB_MESSAGE_HANDLE messageHandle)
+{
+    /* Codes_SRS_IOTHUB_DIAGNOSTIC_13_001: [ IoTHubClient_Diagnostic_AddIfNecessary should return false if diagSetting or messageHandle is NULL. ]*/
+    if (diagSetting == NULL || messageHandle == NULL)
+    {
+        return false;
```

Comment:
the APIs in the C SDK follow a few guidelines that need to be implemented here:
1. Each function shall have only one point of return (create a variable on the top of the function, set it along the function and return its value at the end - see other functions in the C SDK to use as reference);
2. Functions in the C SDK usually return an int (see "__FAILURE__" in other functions in the C SDK)

Reply:
thanks for the detail guidelines